File: library/src/main/java/com/example/library/BaiduLocationService.java
package com.example.library;

import android.content.Context;

import com.baidu.location.BDAbstractLocationListener;
import com.baidu.location.BDLocation;
import com.baidu.location.LocationClient;
import com.baidu.location.LocationClientOption;

public class BaiduLocationService {
    private LocationClient mLocationClient;

    private BaiduLocationListener listener;

    public BaiduLocationService(Context context) {
        //BDAbstractLocationListener为7.2版本新增的Abstract类型的监听接口
        //原有BDLocationListener接口暂时同步保留。具体介绍请参考后文中的说明
        mLocationClient = new LocationClient(context.getApplicationContext());
        //声明LocationClient类

        initOption();

        mLocationClient.registerLocationListener(new MyLocationListener());
        //注册监听函数
    }

    public void setBaiduLocationListener(BaiduLocationListener listener) {
        this.listener = listener;
    }

    public void start() {
        mLocationClient.start();
        //mLocationClient为第二步初始化过的LocationClient对象
        //调用LocationClient的start()方法，便可发起定位请求
    }

    public void stop() {
        mLocationClient.stop();
    }

    private void initOption() {
        LocationClientOption option = new LocationClientOption();

        option.setLocationMode(LocationClientOption.LocationMode.Hight_Accuracy);
        //可选，设置定位模式，默认高精度
        //LocationMode.Hight_Accuracy：高精度；
        //LocationMode. Battery_Saving：低功耗；
        //LocationMode. Device_Sensors：仅使用设备；

        option.setCoorType("bd09ll");
        //可选，设置返回经纬度坐标类型，默认gcj02
        //gcj02：国测局坐标；
        //bd09ll：百度经纬度坐标；
        //bd09：百度墨卡托坐标；
        //海外地区定位，无需设置坐标类型，统一返回wgs84类型坐标

        option.setScanSpan(1000);
        //可选，设置发起定位请求的间隔，int类型，单位ms
        //如果设置为0，则代表单次定位，即仅定位一次，默认为0
        //如果设置非0，需设置1000ms以上才有效

        option.setOpenGps(true);
        //可选，设置是否使用gps，默认false
        //使用高精度和仅用设备两种定位模式的，参数必须设置为true

        option.setLocationNotify(true);
        //可选，设置是否当GPS有效时按照1S/1次频率输出GPS结果，默认false

        option.setIgnoreKillProcess(false);
        //可选，定位SDK内部是一个service，并放到了独立进程。
        //设置是否在stop的时候杀死这个进程，默认（建议）不杀死，即setIgnoreKillProcess(true)

        option.SetIgnoreCacheException(false);
        //可选，设置是否收集Crash信息，默认收集，即参数为false

        option.setWifiCacheTimeOut(5 * 60 * 1000);
        //可选，7.2版本新增能力
        //如果设置了该接口，首次启动定位时，会先判断当前WiFi是否超出有效期，若超出有效期，会先重新扫描WiFi，然后定位

        option.setEnableSimulateGps(false);
        //可选，设置是否需要过滤GPS仿真结果，默认需要，即参数为false

        mLocationClient.setLocOption(option);
        //mLocationClient为第二步初始化过的LocationClient对象
        //需将配置好的LocationClientOption对象，通过setLocOption方法传递给LocationClient对象使用
        //更多LocationClientOption的配置，请参照类参考中LocationClientOption类的详细说明
    }


    public class MyLocationListener extends BDAbstractLocationListener {
        @Override
        public void onReceiveLocation(BDLocation location) {
            //此处的BDLocation为定位结果信息类，通过它的各种get方法可获取定位相关的全部结果
            //以下只列举部分获取经纬度相关（常用）的结果信息
            //更多结果信息获取说明，请参照类参考中BDLocation类中的说明

            double latitude = location.getLatitude();    //获取纬度信息
            double longitude = location.getLongitude();    //获取经度信息
            float radius = location.getRadius();    //获取定位精度，默认值为0.0f

            String coorType = location.getCoorType();
            //获取经纬度坐标类型，以LocationClientOption中设置过的坐标类型为准

            int errorCode = location.getLocType();
            //获取定位类型、定位错误返回码，具体信息可参照类参考中BDLocation类中的说明


            if (listener != null) {
                listener.getLocation(latitude, longitude, radius);
                listener.geterrCodeOrLocType(errorCode);
            }
        }
    }

    /**
     * 百度定位回调接口
     */
    public interface BaiduLocationListener {
        /**
         * @param latitude  纬度
         * @param longitude 经度
         * @param radius    精度
         */
        void getLocation(double latitude, double longitude, float radius);

        /**
         * 获取定位类型、定位错误返回码
         * <p>
         * <pre>
         * 返回值	返回值说明
         * 61	GPS定位结果，GPS定位成功
         * 62	无法获取有效定位依据，定位失败，请检查运营商网络或者WiFi网络是否正常开启，尝试重新请求定位
         * 63	网络异常，没有成功向服务器发起请求，请确认当前测试手机网络是否通畅，尝试重新请求定位
         * 66	离线定位结果。通过requestOfflineLocaiton调用时对应的返回结果
         * 68	网络连接失败时，查找本地离线定位时对应的返回结果
         * 161	网络定位结果，网络定位成功
         * 162	请求串密文解析失败，一般是由于客户端SO文件加载失败造成，请严格参照开发指南或demo开发，放入对应SO文件
         * 167	服务端定位失败，请您检查是否禁用获取位置信息权限，尝试重新请求定位
         * 505	AK不存在或者非法，请按照说明文档重新申请AK
         * </pre>
         *
         * @param errCodeOrLocType
         */
        void geterrCodeOrLocType(int errCodeOrLocType);
    }


}

File: libraries/library/src/main/java/com/example/library/BaiduLocationService.kt
package com.example.library

import android.content.Context
import com.baidu.location.LocationClient
import com.baidu.location.LocationClientOption

/**
 *
 * <pre>
 * 返回值	返回值说明
 * 61	GPS定位结果，GPS定位成功
 * 62	无法获取有效定位依据，定位失败，请检查运营商网络或者WiFi网络是否正常开启，尝试重新请求定位
 * 63	网络异常，没有成功向服务器发起请求，请确认当前测试手机网络是否通畅，尝试重新请求定位
 * 66	离线定位结果。通过requestOfflineLocaiton调用时对应的返回结果
 * 68	网络连接失败时，查找本地离线定位时对应的返回结果
 * 161	网络定位结果，网络定位成功
 * 162	请求串密文解析失败，一般是由于客户端SO文件加载失败造成，请严格参照开发指南或demo开发，放入对应SO文件
 * 167	服务端定位失败，请您检查是否禁用获取位置信息权限，尝试重新请求定位
 * 505	AK不存在或者非法，请按照说明文档重新申请AK
 * </pre>
 * Created by Siy on 2018/5/17.
 *
 * @author Siy
 * @date 2018/5/17.
 */
class BaiduLocationService(context: Context) {
    private val mLocationClient: LocationClient = LocationClient(context)

    var listener: ((Double?, Double?, Float?, Int?) -> Unit)? = null

    init {
        initOption()
        mLocationClient.registerLocationListener { location ->
            listener?.invoke(location?.latitude, location?.longitude, location?.radius, location?.locType)
        }
    }

    fun start() {
        if (!mLocationClient.isStarted) {
            mLocationClient.start()
        }

    }

    fun stop() {
        if (mLocationClient.isStarted) {
            mLocationClient.stop()
        }
    }

    private fun initOption() {
        val option = LocationClientOption()
        option.locationMode = LocationClientOption.LocationMode.Hight_Accuracy
        option.coorType = "bd09ll"
        option.scanSpan = 1000
        option.openGps = true
        option.isLocationNotify = true
        option.isIgnoreKillProcess = false
        option.isIgnoreCacheException = false
        option.wifiCacheTimeOut = 5 * 60 * 1000
        option.enableSimulateGps = false
        mLocationClient.locOption = option
    }
}